### PR TITLE
Add IBAN in our classes and add the jar of the Bankboys

### PR DESCRIPTION
**Pull request by clenrock**
- Relevant Issues: - #241 
- Related Pull Requests: -

* * *

## What

This Pull Request adds to the repository that we can use the IBAN object in the Profile and Contact and send this through Bluetooth. Also the jar build by [BankChain](https://github.com/MatthijsKok/TI2806-Contextproject).

## Why

This Pull Request is needed because we would like to integrate this jar and its functionality into our app.

## How

This feature can be viewed/tested within the project by looking at the Contact page and Settings page.

## Notes

This is a duplicate PR from PR #300 but that one was closed because we will use "integration-develop" for working on the integration.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -82,6 +83,7 @@ dependencies {
     androidTestCompile('com.android.support.test.espresso:espresso-intents:2.2.2', {
         exclude group: 'com.android.support', module: 'support-annotations'
     })
+
```
Comment (ericcornelissen) [inline app/build.gradle]: Can you remove this enter 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/ChangeContactActivity.java [hunk 2]
```diff
@@ -83,6 +101,13 @@ public void saveContact(final View v) {
             setResult(RESULT_FIRST_USER,
                     new Intent().putExtra(ConstantKeywords.CONTACT, contact));
             finish();
+        } else if (validName) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/ChangeContactActivity.java]: If the name is valid, display a warning popup?

File: app/src/main/java/com/nervousfish/nervousfish/activities/ChangeContactActivity.java [hunk 3]
```diff
@@ -67,11 +73,23 @@ public void saveContact(final View v) {
...
+        if (!validIban && (EMPTY_STRING.equals(ibanString) || "-".equals(ibanString))) {
+            validIban = true;
+            ibanString = EMPTY_STRING;
+        }
+        if (validName && validIban) {
             //Update contact
             try {
-                final Contact newContact = new Contact(editText.getText().toString(), contact.getKeys());
+                Contact newContact = new Contact(editTextName.getText().toString(), contact.getKeys());
+                if (!ibanString.equals(EMPTY_STRING)) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/ChangeContactActivity.java]: Judging from the `CreateProfileActivity`, the IBan **is** required. Why is it not required here?

File: app/src/main/java/com/nervousfish/nervousfish/activities/ChangeContactActivity.java [hunk 4]
```diff
@@ -111,8 +136,11 @@ private boolean isValidName(final String name) {
          */
         @Override
         public void onClick(final View v) {
-            final EditText editText = (EditText) findViewById(R.id.edit_contact_name_input);
-            if (editText.getText().toString().equals(contact.getName())) {
+            final EditText editTextName = (EditText) findViewById(R.id.edit_contact_name_input);
+            final EditText editTextIBAN = (EditText) findViewById(R.id.contact_page_change_iban);
+            if (editTextName.getText().toString().equals(contact.getName())
+                    && (editTextIBAN.getText().toString().equals(contact.getIban())
+                    || editTextIBAN.getText().toString().equals(ChangeContactActivity.this.getString(R.string.dash)))) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/ChangeContactActivity.java]: This if-statement is not readable, can you refactor it please 🙂

File: app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileActivity.java [hunk 5]
```diff
@@ -32,15 +35,19 @@
...
  * The {@link android.app.Activity} that is used to create a user profile when the app is first
  * used.
  */
-@SuppressWarnings("checkstyle:ReturnCount")
-//Suppresses return count to allow multiple returncodes while checking input fields.
+@SuppressWarnings({"checkstyle:ReturnCount", "PMD.CyclomaticComplexity", "PMD.StdCyclomaticComplexity",
+    "PMD.ExcessiveImports"})
+//1. Suppresses return count to allow multiple returncodes while checking input fields.
+//2 and 3. The complexity does not make the code unreadible at this moment.
+//4. We want to have so much imports so we dont have to write Constants.ExplicitFieldResultCodes every time.
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileActivity.java]: Not sure if this is a good reason to suppress "ExcessiveImports" 🤔

File: app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java [hunk 6]
```diff
@@ -79,6 +83,22 @@
 
     /**
      * @param input The {@link EditText} to evaluate.
+     * @return The resultcode of validateIban.
+     */
+    Constants.InputFieldResultCodes validateIban(final EditText input) {
+        final String iban = input.getText().toString();
+        if (EMPTY_STRING.equals(iban) || IBANVerifier.isValidIBAN(iban)) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java]: Doesn't `IBANVerifier.isValidIBAN(iban)` return false if `iban` is an empty string?

File: app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java [hunk 7]
```diff
@@ -109,11 +112,19 @@ public void onActivityResult(final int requestCode, final int resultCode, final
...
+            final IKey key = QRGenerator.deconstructToKey(result.split(SEMI_COLON)[2]);
 
             final IDatabase database = this.serviceLocator.getDatabase();
-            final Contact contact = new Contact(name, key);
+            Contact contact;
+            try {
+                contact = new Contact(name, key, new IBAN(iban));
+            } catch (final IllegalArgumentException e) {
+                LOGGER.info("IBAN is not valid", e);
+                contact = new Contact(name, key);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java]: Do we even want this functionality, given that the IBan is required in the `CreateProfileActivity`?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 8]
```diff
@@ -41,7 +45,8 @@
 public final class SettingsActivity extends AAppCompatPreferenceActivity {
 
     private static final Logger LOGGER = LoggerFactory.getLogger("SettingsActivity");
-    private static boolean firstLoad = true;
+    private static boolean firstLoadName = true;
+    private static boolean firstLoadIban = true;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Why are these attributes useful? 🤔

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java [hunk 9]
```diff
@@ -11,15 +11,22 @@
...
+import java.util.Objects;
+
+import nl.tudelft.ewi.ds.bankver.IBAN;
 
 /**
  * A Java object representation (POJO) of a contact.
  */
+@SuppressWarnings({"PMD.UselessParentheses", "PMD.NullAssignment"})
+//1. The parentheses on line 125 are not useless.
+//2. We want iban to be immutable, but not a required field.
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: Are you sure? The IBan seems to be a required field in the `CreateProfileActivity`...

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Profile.java [hunk 10]
```diff
@@ -8,15 +8,22 @@
...
+
+import nl.tudelft.ewi.ds.bankver.IBAN;
 
 /**
  * A Profile POJO to store a contact representing the user and the user's keypairs.
  */
-public class Profile implements Serializable {
+@SuppressWarnings({"PMD.UselessParentheses", "PMD.NullAssignment"})
+//1. The parentheses on line 104 are not useless.
+//2. We want iban to be immutable, but not a required field.
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Profile.java]: Are you sure? The IBan seems to be a required field in the `CreateProfileActivity`...

File: app/src/main/res/layout/activity_change_contact.xml [hunk 11]
```diff
@@ -6,7 +6,8 @@
     android:layout_height="match_parent"
     android:fitsSystemWindows="true"
     android:orientation="vertical"
-    tools:context="com.nervousfish.nervousfish.activities.ContactActivity">
+    tools:context="com.nervousfish.nervousfish.activities.ContactActivity"
+    tools:ignore="LabelFor">
```
Comment (ericcornelissen) [inline app/src/main/res/layout/activity_change_contact.xml]: Why the ignore? 🤔

File: app/src/main/res/layout/activity_change_contact.xml [hunk 12]
```diff
@@ -88,15 +89,15 @@
...
+                        android:contentDescription="@string/contact_image"
+                        android:src="@drawable/personplaceholder" />
 
                     <EditText
-                        android:labelFor="@id/iv_change_contact"
                         android:id="@+id/edit_contact_name_input"
                         android:layout_width="wrap_content"
                         android:layout_height="match_parent"
                         android:inputType="text"
+                        android:labelFor="@id/iv_change_contact"
```
Comment (ericcornelissen) [inline app/src/main/res/layout/activity_change_contact.xml]: Could you remove the ignore from this element 😇

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -55,6 +56,11 @@
     <string name="create_profile_error_adding_to_database">Error while adding profile to database</string>
     <string name="create_profile_passwords_not_equal">Please make sure you typed the same password twice.</string>
     <string name="create_profile_all_fields_empty">Please fill in the fields</string>
+    <string name="create_profile_invalid_iban">The IBAN number is invalid, please enter a valid one.</string>
+
+    <string name="name_create_profile">Name*</string>
+    <string name="repeat_password">Repeat password*</string>
+    <string name="password_create_profile">Password*</string>
```
Comment (ericcornelissen) [inline app/src/main/res/values/strings.xml]: Why did you change these string names/Why are these not listed under `<!-- Common sentences -->` or `<!-- Words -->`

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -80,6 +86,10 @@
     <string name="contact_deleted_description">The contact has been deleted!</string>
     <string name="contact_doesnt_exist">The contact doesn\'t exist</string>
     <string name="contact_already_deleted">It looks like the contact was already deleted.</string>
+    <string name="contact_iban">IBAN</string>
```
Comment (ericcornelissen) [inline app/src/main/res/values/strings.xml]: This is to generic to be listed under `<!-- Contacts -->`, can you list it under `<!-- Words -->` and change its name to "iban"

File: config/quality/checkstyle/checkstyle.xml [hunk 15]
```diff
@@ -48,7 +48,9 @@
         <module name="DeclarationOrder" />
         <module name="DefaultComesLast" />
         <module name="EmptyStatement" />
-        <module name="EqualsAvoidNull" />
+        <module name="EqualsAvoidNull" >
+            <property name="severity" value="ignore" />
```
Comment (ericcornelissen) [inline config/quality/checkstyle/checkstyle.xml]: Can you remove this ignore and suppress the issue where you need it 😇 I do (generally) want to avoid this...

File: app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java [hunk 16]
```diff
@@ -79,6 +83,22 @@
 
     /**
      * @param input The {@link EditText} to evaluate.
+     * @return The resultcode of validateIban.
+     */
+    Constants.InputFieldResultCodes validateIban(final EditText input) {
+        final String iban = input.getText().toString();
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java]: When you closed an earlier PR of this branch you commented on my Validate suggestions if we were already done discussing validations. I think that we should validate our user input, not just because we got whole lectures about checking the class invariant, pre-conditions, post-conditions, etc. but also because it gives you much more safety because you're sure that your input meets certain requirements (i.e. it's literally impossible to get nullPointerExceptions if you validate your input)

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java]: So, yeah...
`Validate.notNull(input);`

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java [hunk 17]
```diff
@@ -30,6 +37,21 @@
     public Contact(final String name, final IKey key) {
         this.name = name;
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]:
```
Validate.notBlank(name);
Validate.notNull(key);
```

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java [hunk 18]
```diff
@@ -30,6 +37,21 @@
...
+    /**
+     * Constructor for the {@link Contact} POJO for a single {@link IKey},
+     * and an IBAN.
+     *
+     * @param name The name of the {@link Contact}
+     * @param key  The {@link IKey} to initialize the {@link Contact} with
+     * @param iban  The IBAN of the {@link Contact}
+     */
+    public Contact(final String name, final IKey key, final IBAN iban) {
+        this.name = name;
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]:
```
Validate.notBlank(name);
Validate.notNull(key);
```

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java [hunk 19]
```diff
@@ -43,12 +65,45 @@ public Contact(final String name, final Collection<IKey> keys) {
...
+    /**
+     * Constructor for the {@link Contact} POJO for multiple {@link IKey}s, and
+     * the IBAN.
+     *
+     * @param name The name of the {@link Contact}
+     * @param keys The {@link Collection} of keys to initialize the {@link Contact} with
+     * @param iban  The IBAN of the {@link Contact}
+     */
+    public Contact(final String name, final Collection<IKey> keys, final IBAN iban) {
+        this.name = name;
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]:
```
Validate.notBlank(name);
Validate.noNullElements(key);
```

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Profile.java [hunk 20]
```diff
@@ -28,6 +35,21 @@
     public Profile(final String name, final List<KeyPair> keyPairs) {
         this.keyPairs = keyPairs;
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Profile.java]:
```
Validate.notBlank(name);
Validate.noNullElements(keyPairs);
```

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 21]
```diff
@@ -104,6 +113,50 @@ private void updateDisplayName(final Preference preference, final String stringV
...
 
         /**
+         * If the key is iban_number this method is called to update the summary
+         * and the Profile in the database.
+         *
+         * @param preference The preference which is changed
+         * @param stringValue The string value which is new
+         */
+        private void updateIban(final Preference preference, final String stringValue) {
+            if (firstLoadIban) {
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: assert prefence != null;
assert stringValue != null;
assert stringValue.size() == 18;

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (clenrock) [inline app/src/main/res/values/strings.xml]: These are the extra strings which are only used in the CreateProfileActivity to show that those are required

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java (hunk 8 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: For loading the first time, we should get the data from the database. Otherwise we could get them from the filled in preference

File: app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileActivity.java (hunk 5 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileActivity.java]: The exports are all constants from "Constants.ExplicitFieldResultCodes"

File: app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java (hunk 7 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java]: IBAN is not required in the CreateProfileActivity

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java (hunk 9 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: IBAN is not required in the CreateProfileActivity

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Profile.java (hunk 10 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Profile.java]: IBAN is not required in the CreateProfileActivity

File: app/src/main/res/layout/activity_change_contact.xml (hunk 11 above)
Comment (clenrock) [inline app/src/main/res/layout/activity_change_contact.xml]: Lint says that every Edittext should have another element pointing to it with Labelfor, but here that is not suited.
